src/util/readwritefile.cpp

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2015-2020 The Bitcoin Core developers
+// Copyright (c) 2017 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <fs.h>
+
+#include <limits>
+#include <stdio.h>
+#include <string>
+#include <utility>
+
+std::pair<bool,std::string> ReadBinaryFile(const fs::path &filename, size_t maxsize=std::numeric_limits<size_t>::max())
+{
+    FILE *f = fsbridge::fopen(filename, "rb");
+    if (f == nullptr)
+        return std::make_pair(false,"");
+    std::string retval;
+    char buffer[128];
+    do {
+        const size_t n = fread(buffer, 1, sizeof(buffer), f);
+        // Check for reading errors so we don't return any data if we couldn't
+        // read the entire file (or up to maxsize)
+        if (ferror(f)) {
+            fclose(f);
+            return std::make_pair(false,"");
+        }
+        retval.append(buffer, buffer+n);
+    } while (!feof(f) && retval.size() <= maxsize);
```

Comment:
Yes, it should get out of the loop and it does due to `retval.size() <= maxsize`.

It is true that if `maxsize==10` and `fread()` returns 7 bytes and then 5 bytes, this function will return a `retval` that contains 12 bytes, exceeding `maxsize`. This is also how it works in `master` right now.

For the purposes of this PR it suffices to move `ReadBinaryFile()` out of `torcontrol.cpp` so that it can be reused by other code. So I tried to keep changes to the minimum - moved the function and only fixed a gross bug (commit `util: fix ReadBinaryFile() returning partial contents`).